src/export_json.lean

```diff
@@ -340,20 +340,52 @@ do docs ← olean_doc_strings,
      (filename, json.array $ l.map write_module_doc_pair) :: rest
    end) [])
 
-meta def get_instances : tactic (rb_lmap string string) :=
+/-- Extract `[foo, bar]` from `has_pow foo bar`.
+
+The result is a map containing the type names as keys.  -/
+meta def find_instance_types (e : expr) : tactic (rb_map string unit) :=
+list.mfoldl (λ m e, do
+  e ← dunfold [`decidable_rel, `decidable_pred] e {fail_if_unchanged := ff},
+  e ← whnf e transparency.reducible,
+  t ← infer_type e,
+  -- only look at arguments which are types or propositions
+  expr.sort l ← whnf t transparency.reducible | pure m,
+  some s ← match e.get_app_fn with
+  | expr.const type_name l   := pure (some type_name.to_string)
+  | expr.pi _ _ _ _          := pure (some "pi")
+  | expr.sort level.zero     := pure (some "Prop")
+  | expr.sort (level.succ l) := pure (some "Type")
+  | expr.sort l              := pure (some "Sort")
+  | expr.local_const _ _ _ _ := pure none
+  | expr.macro _ _           := pure none -- TODO: unfold macros?
+  | expr.lam _ _ _ _         := fail ("lam, not a constant: " ++ to_string e)
+  | expr.var i               := fail ("var, not a constant: " ++ to_string e)
+  | expr.mvar _ _ _          := fail ("mvar, not a constant: " ++ to_string e)
+  | expr.app _ _             := fail ("app, not a constant: " ++ to_string e)
+  | expr.elet _ _ _ _        := fail ("elet, not a constant: " ++ to_string e)
```

Comment:
This bails out if any argument is a lambda, but we probably just want to ignore the lambda arguments.

Reply:
This is an example where you have a lambda as argument: https://leanprover-community.github.io/mathlib_docs_demo/algebra/graded_monoid.html#set_like.ghas_one

But I guess we skip this argument because it's not a type.

> bailing out

I didn't read the code at first, but apparently the fail is fatal.  (I blindly assumed there was a `<|> pure []` or something afterwards.)  I'd really like to avoid such fatal errors.  I'm already dreading the email from github actions that says "doc-gen failed, elet, not a constant: ...".

Reply:
I added some handling of lambdas